### PR TITLE
feat: adds unit support to ceil, floor, and fix functions

### PR DESCRIPTION
**Pull request by orelbn**
# Description
- adds unit support to ceil, floor, and fix functions -> Addresses #3216

## Test
- Added automated unit-tests

### Additional Notes
- I wasn't able to run the type tests locally. Can you please verify that they are passing properly?

### CODE REVIEW

File: src/function/arithmetic/ceil.js [hunk 1]
```diff
@@ -122,6 +128,24 @@ export const createCeil = /* #__PURE__ */ factory(name, dependencies, ({ typed,
       return x.ceil(n.toNumber())
     },
 
+    'Unit, number, Unit': typed.referToSelf(self => function (x, n, unit) {
```
Comment (josdejong) [inline src/function/arithmetic/ceil.js:131]: I think the code can be simplified from 6 new signatures to 3:

```js
'Unit, number | BigNumber, Unit': ...,
'Array | Matrix, number | BigNumber, Unit': ...,
'Array | Matrix | Unit, Unit': ...
```

And the same for functions `fix` and `floor`. Can you give that a try?

**Review by josdejong — CHANGES_REQUESTED** [review]
Thanks Orel! This is exactly what we're looking for, nice job. Thanks for updating all the docs and TypeScript types. The types run fine, they run via Github Actions of this PR and all checks are green.

I've made one inline comment, can you have a look at that?

**Comment by orelbn** [comment]
Is there a restriction that the matrix or array must contain only units? I believe that to be the case. If it is so I might want to update the types further.

File: src/function/arithmetic/ceil.js [hunk 2]
```diff
@@ -122,6 +128,19 @@ export const createCeil = /* #__PURE__ */ factory(name, dependencies, ({ typed,
       return x.ceil(n.toNumber())
     },
 
+    'Unit, number | BigNumber, Unit': typed.referToSelf(self => function (x, n, unit) {
+      if (typeof n !== 'number') n = n.toNumber()
```
Comment (josdejong) [inline src/function/arithmetic/ceil.js:132]: Can you remove the line `if (typeof n !== 'number') n = n.toNumber()` from each of the functions? We're calling `self(valueless, n)`, and that will do the checking on whether `n` is a number or BigNumber itself :)

File: types/index.d.ts [hunk 3]
```diff
@@ -4995,6 +5007,25 @@ export interface MathJsChain<TValue> {
...
+  ceil(this: MathJsChain<Unit>, unit: Unit): MathJsChain<Unit>
+  ceil<U extends MathCollection>(
+    this: MathJsChain<U>,
+    unit: Unit
+  ): MathJsChain<U>
+  ceil(
+    this: MathJsChain<Unit>,
+    n: number | BigNumber,
+    unit: Unit
+  ): MathJsChain<Unit>
```
Comment (josdejong) [inline types/index.d.ts:5023]: I think `ceil(Unit, number | BigNumber, Unit)` returns a `number | BigNumber` and not a `Unit`, right?

**Review by josdejong — CHANGES_REQUESTED** [review]
Thanks for the updates. I made two more comments, can you have a look at those?

> Is there a restriction that the matrix or array must contain only units? 

Matrices can contain mixed content of numbers, units, complex numbers etc. There is still an issue in that regard though: the type definitions currently do not allow to have a matrix containing units at all: 

https://github.com/josdejong/mathjs/blob/fb76ac419d680543d2ecff70a594e177f3e41b5b/types/index.d.ts#L16

Maybe best to address that issue separately though.

File: src/function/arithmetic/ceil.js (hunk 2 above)
Comment (orelbn) [reply src/function/arithmetic/ceil.js:132]: Got it thanks!

File: src/function/arithmetic/round.js [hunk 4]
```diff
@@ -147,23 +147,18 @@ export const createRound = /* #__PURE__ */ factory(name, dependencies, ({ typed,
...
     },
 
-    'Unit, number, Unit': typed.referToSelf(self => function (x, n, unit) {
+    'Unit, number | BigNumber, Unit': typed.referToSelf(self => function (x, n, unit) {
+      if (typeof n !== 'number') n = n.toNumber()
       const valueless = x.toNumeric(unit)
       return unit.multiply(self(valueless, n))
     }),
 
-    'Unit, BigNumber, Unit': typed.referToSelf(self => (x, n, unit) => self(x, n.toNumber(), unit)),
```
Comment (orelbn) [inline src/function/arithmetic/round.js:155]: It converted it to a number previously.  If there was BigNumber

File: types/index.d.ts (hunk 3 above)
Comment (orelbn) [reply types/index.d.ts:5023]: I use the same type definition as round, which I did not update.
```
  round(
    this: MathJsChain<Unit>,
    n: number | BigNumber,
    unit: Unit
  ): MathJsChain<Unit>
  ```

**Comment by orelbn** [comment]
> Thanks for the updates. I made two more comments, can you have a look at those?
> 
> > Is there a restriction that the matrix or array must contain only units?
> 
> Matrices can contain mixed content of numbers, units, complex numbers etc. There is still an issue in that regard though: the type definitions currently do not allow to have a matrix containing units at all:
> 
> https://github.com/josdejong/mathjs/blob/fb76ac419d680543d2ecff70a594e177f3e41b5b/types/index.d.ts#L16
> 
> Maybe best to address that issue separately though.

Thank you for the clarification; I should have been more specific. My apologies 🙏 . Is there a restriction that matrices and arrays contain only units when the second or third argument to the `ceil, floor, fix, or round` functions is a unit?

For example, `round([1], unit('cm'))` will throw an error and also would `round([1, unit('2 cm')], 2, unit('cm'))`

Therefore, the type definition would be:

`round(Unit[] | Unit[][] | UnitMatrix, number | BigNumber, Unit)`
`round(Unit[] | Unit[][] | UnitMatrix, Unit)`

File: types/index.d.ts (hunk 3 above)
Comment (josdejong) [reply types/index.d.ts:5023]: Ah, sorry my bad. I was mistaken. The functions indeed return a `Unit`.

File: src/function/arithmetic/ceil.js (hunk 2 above)
Comment (josdejong) [reply src/function/arithmetic/ceil.js:132]: Closing this comment in favor of https://github.com/josdejong/mathjs/pull/3269#discussion_r1779340372

File: src/function/arithmetic/round.js (hunk 4 above)
Comment (josdejong) [reply src/function/arithmetic/round.js:155]: When removing the code `if (typeof n !== 'number') n = n.toNumber()` from the functions, all works except that the unit test that you refer to fails. 

The reason for this is the following: in general, when you use mixed numbers and BigNumbers in mathjs functions, the output will become a BigNumber (since a `BigNumber` can hold any `number` so that is a safe conversion, but a `number` cannot hold any `BigNumber`). The failing unit tests has a mix of `number` and `BigNumber` in its arguments. So it is fine when the output contains a BigNumber then: `unit(bignumber('3.12'), 'cm')` instead of `unit('3.12 cm')`. In the case of these rounding functions it may feel a bit odd, since the number of digits `n` is not the "main" value. This is an edge case though and if people really provide mixed `number` and `BigNumber` this way, `BigNumber` output is what they'll get ;). Does that sort of make sense?

In short: can you remove the lines `if (typeof n !== 'number') n = n.toNumber()` and adjust the concerning unit tests to expect a Unit with BigNumber value as output?

Comment (josdejong) [reply src/function/arithmetic/round.js:155]: Hmm. You are right though about that this behavior is already there in function `round`, so changing it would be a breaking change 🤔 . That is a pity. 

Thinking aloud...

Let's not introduce a breaking change here, and let's keep the behavior of function `round` as it is, and align the other functions with that (like you already have done right now).

So then there are two options I think:

1. have unified signatures like `Unit, number | BigNumber, Unit` and special logic `if (typeof n !== 'number') n = n.toNumber()`.
2. have separate signatures like `Unit, number, Unit` and `Unit, BigNumber, Unit`.

I think from these two options I like 2 best (your original approach): I prefer not having `typeof` checks in the logic and let all of that be handled via the function signatures. What do you think?


Sorry about the confusion and all the back-and-forth, I hand't realized that much of the behavior is already in function `round` 😕.

**Comment by josdejong** [comment]
> Is the desired behaviour that matrices and arrays only contain units when the second or third argument to the `ceil, floor, fix, or round` function is a unit?

That makes sense indeed! Can you refine the type definitions with your proposal?

**Comment by orelbn** [comment]
> > Is the desired behaviour that matrices and arrays only contain units when the second or third argument to the `ceil, floor, fix, or round` function is a unit?
> 
> That makes sense indeed! Can you refine the type definitions with your proposal?

Sorry to leave you hanging; very busy this upcoming week. So probably won't get to this until October 12th.

All the comments make sense.

File: src/function/arithmetic/round.js (hunk 4 above)
Comment (orelbn) [reply src/function/arithmetic/round.js:155]: Will go with option 2.

**Comment by orelbn** [comment]
> > Is the desired behaviour that matrices and arrays only contain units when the second or third argument to the `ceil, floor, fix, or round` function is a unit?
> 
> That makes sense indeed! Can you refine the type definitions with your proposal?

I played around with it a bit, and I think it would be better to do it in a separate PR. 

This is what I believe the tasks that are required:

1. Expand the Matrix and MathArray types to accept Units.
2. Add a utility for checking whether an Array  is a UnitArray and Matrix is a UnitMatrix

From a user perspective, if they create an array that contains multiple types they will get this error
```
  const array = []
  array.push(u1)
  array.push(u3)
  array.push(1)
  array.pop()

  math.ceil(array, 1, math.unit('cm'))
```
Pasted Graphic

They can use the utility (although slow since it has to go through. the entire array/matrix) or assert the type.

 `if (isUnitArray(array)) math.ceil(array, 1, math.unit('cm'))`
or
`math.ceil(array as UnitArray, 1, math.unit('cm'))`

File: src/function/arithmetic/round.js (hunk 4 above)
Comment (josdejong) [reply src/function/arithmetic/round.js:155]: Thanks, looks good. Nice to at least have 4 signatures instead of originally 6.

**Comment by josdejong** [comment]
Thanks for the updates!

The improved TypeScript support is awesome, and with the generic `MathCollection<T>` we can refine a lot more of the type definitions. However we're already getting out of scope of this PR so let's limit what we do in this PR 😅.

Two questions:

1. There are some subtle changes in the existing type definitions (like for `multiply`). Are you sure there are no breaking changes in the types?
2. Thinking aloud here: I'm a bit in doubt about the typeguards `isUnitArray` and `isUnitMatrix` as they are. I think we can use more fundamental helper functions to determine whether a matrix/array contains a single data type (i.e. is homogeneous). Not just for units but for any type. So we would need something more generic like `isHomogeneousArray(array, isComplex)` or `isArrayOfType(isComplex)`. The ideal TypeScript notation would be something like `isArray<Complex>()`. Maybe we should implement it a method for this on `DenseMatrix` and `SparseMatrix` which verifies the contents and sets the internal `datatype` (which optimizes evaluation) and sets the TypeScript type. 

    In short: this needs some more thinking-through, and maybe it is best to keep that out of scope of this PR and address in a separate discussion. Until then, it would be best not to export new helper functions like `isUnitArray` as long as we're not sure what the final API will be. What do you think?

**Comment by orelbn** [comment]
> Thanks for the updates!
> 
> The improved TypeScript support is awesome, and with the generic `MathCollection<T>` we can refine a lot more of the type definitions. However we're already getting out of scope of this PR so let's limit what we do in this PR 😅.
> 
> Two questions:
> 
> 1. There are some subtle changes in the existing type definitions (like for `multiply`). Are you sure there are no breaking changes in the types?
> 2. Thinking aloud here: I'm a bit in doubt about the typeguards `isUnitArray` and `isUnitMatrix` as they are. I think we can use more fundamental helper functions to determine whether a matrix/array contains a single data type (i.e. is homogeneous). Not just for units but for any type. So we would need something more generic like `isHomogeneousArray(array, isComplex)` or `isArrayOfType(isComplex)`. The ideal TypeScript notation would be something like `isArray<Complex>()`. Maybe we should implement it a method for this on `DenseMatrix` and `SparseMatrix` which verifies the contents and sets the internal `datatype` (which optimizes evaluation) and sets the TypeScript type.
>    In short: this needs some more thinking-through, and maybe it is best to keep that out of scope of this PR and address in a separate discussion. Until then, it would be best not to export new helper functions like `isUnitArray` as long as we're not sure what the final API will be. What do you think?

Sorry for the delayed response. Very busy this past week.

All of your suggestions sound great. Thank you for the input! I will remove the type guards and address your suggestion in a separate PR. I agree that there is definitely more thinking to do. 

Regarding the multiply, I will take a look again (I added a test, but I need to place it somewhere else). There were some types that I believe were inaccurate previously. For example, if you multiplied two arrays, it assumed the result was an array.

https://github.com/josdejong/mathjs/blob/07bff5501dae43c1243afc2d9c4b81f70e807098/types/index.d.ts#L1364C3-L1364C47

However, from my testing in the playground, I found this was not the case.

image

I am worried about breaking changes. I do not have enough context into the whole repo to be aware of breaking changes, so I am relying on previous unit tests as a check.

**Comment by orelbn** [comment]
I did find a small error in one of the multiply function types. Would you like me to file an issue or fix it in one of the upcoming PRs?

Error in function type: multiplication of two MathArrays does not necessarily result in a MathArray.

https://github.com/josdejong/mathjs/blob/07bff5501dae43c1243afc2d9c4b81f70e807098/types/index.d.ts#L1364

**To Reproduce**
Steps to reproduce the behavior.

Multiply two single-dimensional arrays together, and the result is a number.

image
image
image

File: test/typescript-tests/testTypes.ts [hunk 5]
```diff
@@ -2371,9 +2484,6 @@ Factory Test
   if (math.isMatrix(x)) {
     expectTypeOf(x).toMatchTypeOf<Matrix>()
   }
-  if (math.isDenseMatrix(x)) {
```
Comment (josdejong) [inline test/typescript-tests/testTypes.ts:2374]: I think it would be good to leave this `math.isDenseMatrix` test in place? Is there a special reason to remove it?

**Review by josdejong — CHANGES_REQUESTED** [review]
Thanks for the updates, I think we're there! I made one last small remark, can you have a look at that?

Good find about the wrong types when multiplying two vectors. In that case mathjs returns a scalar value, and not an array containing one value. Since this is not a regression introduced in this PR but an existing bug in the type definitions, I think it's best to create a separate PR improving the types for vector multiplication, ok?

I've had a thorough look at the changes in the type definitions. I expect the changes are not be breaking in normal, practical cases, but there may be tricky edge cases that have breaking behavior. To be careful in this regard, I would like to merge this PR into the next major version v14, ok? (I plan to release v14 "soonish", within a couple of weeks).

File: test/typescript-tests/testTypes.ts (hunk 5 above)
Comment (orelbn) [reply test/typescript-tests/testTypes.ts]: My apologizes removed that by accident, I added it back.

**Comment by orelbn** [comment]
> ad a thorough look at the changes in the type definitions. I expect the changes are not be breaking in normal, practical cases, but there may be tricky edge cases that have breaking behavior. To be careful in this reg

That sounds good. I will address the vector multiplication types separately.

Adding the functionality in version 14 is completely fine with me.

Thank you for the in-depth review 😄

**Review by josdejong — APPROVED** [review]

**Comment by josdejong** [comment]
Thanks! Merged now in the `v14` branch.

**Comment by josdejong** [comment]
Published now in `v14.0.0`. Thanks again Orel.
